lib/icinga/dependency.cpp

```diff
@@ -9,106 +9,170 @@
 #include "base/exception.hpp"
 #include <map>
 #include <sstream>
+#include <unordered_map>
 #include <utility>
+#include <variant>
 
 using namespace icinga;
 
 REGISTER_TYPE(Dependency);
 
-bool Dependency::m_AssertNoCyclesForIndividualDeps = false;
+INITIALIZE_ONCE(Dependency::StaticInitialize);
 
-struct DependencyCycleNode
+void Dependency::StaticInitialize()
 {
-	bool Visited = false;
-	bool OnStack = false;
-};
-
-struct DependencyStackFrame
-{
-	ConfigObject::Ptr Node;
-	bool Implicit;
-
-	inline DependencyStackFrame(ConfigObject::Ptr node, bool implicit = false) : Node(std::move(node)), Implicit(implicit)
-	{ }
-};
-
-struct DependencyCycleGraph
-{
-	std::map<Checkable::Ptr, DependencyCycleNode> Nodes;
-	std::vector<DependencyStackFrame> Stack;
-};
-
-static void AssertNoDependencyCycle(const Checkable::Ptr& checkable, DependencyCycleGraph& graph, bool implicit = false);
+	dynamic_cast<ConfigType*>(Dependency::TypeInstance.get())->BeforeOnAllConfigLoaded.connect(BeforeOnAllConfigLoadedHandler);
+}
 
-static void AssertNoParentDependencyCycle(const Checkable::Ptr& parent, DependencyCycleGraph& graph, bool implicit)
+/**
+ * Helper class to search for cycles in the dependency graph.
+ *
+ * State is stored inside the class and no synchronization is done,
+ * hence instances of this class must not be used concurrently.
+ */
+class DependencyCycleChecker
 {
-	if (graph.Nodes[parent].OnStack) {
-		std::ostringstream oss;
-		oss << "Dependency cycle:\n";
-
-		for (auto& frame : graph.Stack) {
-			oss << frame.Node->GetReflectionType()->GetName() << " '" << frame.Node->GetName() << "'";
+	struct Node
+	{
+		bool Visited = false;
+		bool OnStack = false;
+		std::vector<Dependency::Ptr> ExtraDependencies;
+	};
+
+	// std::hash is only implemented starting from Boost 1.74, use a (ordered) map as fallback.
+	// https://github.com/boostorg/smart_ptr/commit/5a18ffdc5609a0e64b63e47cb81c4f0847e0c087
+#if BOOST_VERSION >= 107400
+	std::unordered_map<Checkable::Ptr, Node> m_Nodes;
+#else
+	std::map<Checkable::Ptr, Node> m_Nodes;
+#endif
+
+	// Stack representing the path currently visited by AssertNoCycle(). Dependency::Ptr represents an edge from its
+	// child to parent, Service::Ptr represents the implicit dependency of that service to its host.
+	std::vector<std::variant<Dependency::Ptr, Service::Ptr>> m_Stack;
+
+public:
+	/**
+	 * Add a dependency to this DependencyCycleChecker that will be considered by AssertNoCycle() in addition to
+	 * dependencies already registered to the checkables. This allows checking if additional dependencies would cause
+	 * a cycle before actually registering them to the checkables.
+	 *
+	 * @param dependency Dependency to additionally consider during the cycle search.
+	 */
+	void AddExtraDependency(Dependency::Ptr dependency)
+	{
+		auto& node = m_Nodes[dependency->GetChild()];
+		node.ExtraDependencies.push_back(std::move(dependency));
```

Comment:
First copying and moving `dependency` to the container should actually make no difference to changing the parameter to a const ref; in the latter case you would actually save the unnecessary move operation here.

Reply:
Taking the argument by value is a trick to get some okay behavior both for moving and copying without having to provide multiple overloads for the cost of an additional move (which is a trivial operation for `intrusive_ptr`).

However, for this to make some sense, I'd actually actually have to use it properly:

https://github.com/Icinga/icinga2/blob/86ba25083d8028704e059100f70c7973e6b51af4/lib/icinga/dependency.cpp#L163-L168

Changing that to `checker.AddExtraDependency(std::move(dependency));` will actually move (possibly twice if not optimized out) compared to a copy then destroying the local variable when the argument type was changed to `const Dependency::Ptr&`.